sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanner.scala

```diff
@@ -47,7 +47,14 @@ class SparkPlanner(
       Window ::
       JoinSelection ::
       InMemoryScans ::
-      BasicOperators :: Nil)
+      BasicOperators(queryExecutionThreadLocal.get()) :: Nil)
+
+  private val queryExecutionThreadLocal = new ThreadLocal[QueryExecution]
```

Comment:
why do we need this?

Reply:
You know, my goal was to pass the current `QueryExecution` instance around to `RecursiveRelationExec` for reporting metrics updates.
- My first attempt was to change the `def apply(plan: LogicalPlan): Seq[SparkPlan]` method to accept some other parameters besides the `plan`. I tried changing `GenericStrategy`'s `apply()` to something like `def apply(plan: LogicalPlan, param: PlanParam): Seq[PhysicalPlan]` where `PlanParam` would be `QueryExecution`. I felt that such a big change would not be acceptable, but here you can find this approach: https://github.com/peter-toth/spark/commit/3a3ac462bb350e1103d640e83394f65994c5175e
- Then my second try was to create new instance of `SparkPlanner` for each `QueryExecution` and I could then pass the current `QueryExecution` to `SparkPlanner` and then to `BasicOperators` and then to `RecursiveRelationExec`. (`IncrementalExecution` does something similar.) But then `SessionState.planner` would became almost useless. I also wasn't sure that having as many planners as `QueryExecution`s would be a good idea.
- Then this current thread-safe approach with a `ThreadLocal` came across my mind, because it requires minimal changes.

Please let me know if you have any better idea. Suggestions are very welcome.

Reply:
I am wondering whether we can avoid to pass the QueryExecution at all...I see it is needed for sending events...I am wondering whether we can leave that out too for the moment...

Reply:
Sure, we can remove this part, especially if it makes the PR unacceptable.

@maropu @cloud-fan @maryannxue @gatorsmile do you have any comments on this or other parts of the PR?